library/std/src/sys/pal/unix/futex.rs

```diff
@@ -135,6 +139,186 @@ pub fn futex_wake_all(futex: &AtomicU32) {
     };
 }
 
+/// With macOS version 14.4, Apple introduced a public futex API. Unfortunately,
+/// our minimum supported version is 10.12, so we need a fallback API. Luckily
+/// for us, the underlying syscalls have been available since exactly that
+/// version, so we just use those when needed. This is private API however,
+/// which means we need to take care to avoid breakage if the syscall is removed
+/// and to avoid apps being rejected from the App Store. To do this, we use weak
+/// linkage emulation for both the public and the private API.
+///
+/// See os/os_sync_wait_on_address.h (Apple has failed to upload the documentation
+/// to their website) for documentation of the public API and
+/// https://github.com/apple-oss-distributions/xnu/blob/1031c584a5e37aff177559b9f69dbd3c8c3fd30a/bsd/sys/ulock.h#L69
+/// for the header file of the private API, along with its usage in libpthread
+/// https://github.com/apple-oss-distributions/libpthread/blob/d8c4e3c212553d3e0f5d76bb7d45a8acd61302dc/src/pthread_cond.c#L463
+#[cfg(any(target_os = "macos", target_os = "ios", target_os = "tvos", target_os = "watchos",))]
+mod apple {
+    use crate::ffi::{c_int, c_void};
+    use crate::sys::pal::weak::weak;
+
+    pub const OS_CLOCK_MACH_ABSOLUTE_TIME: u32 = 32;
+    pub const OS_SYNC_WAIT_ON_ADDRESS_NONE: u32 = 0;
+    pub const OS_SYNC_WAKE_BY_ADDRESS_NONE: u32 = 0;
+
+    pub const UL_COMPARE_AND_WAIT: u32 = 1;
+    pub const ULF_WAKE_ALL: u32 = 0x100;
+    // The syscalls support directly returning errors instead of going through errno.
+    pub const ULF_NO_ERRNO: u32 = 0x1000000;
+
+    // These functions appeared with macOS 14.4, iOS 17.4, tvOS 17.4, watchOS 10.4, visionOS 1.1.
+    weak! {
+        pub fn os_sync_wait_on_address(*mut c_void, u64, usize, u32) -> c_int
+    }
+
+    weak! {
+        pub fn os_sync_wait_on_address_with_timeout(*mut c_void, u64, usize, u32, u32, u64) -> c_int
+    }
+
+    weak! {
+        pub fn os_sync_wake_by_address_any(*mut c_void, usize, u32) -> c_int
+    }
+
+    weak! {
+        pub fn os_sync_wake_by_address_all(*mut c_void, usize, u32) -> c_int
+    }
+
+    // This syscall appeared with macOS 11.0.
+    // It is used to support nanosecond precision for timeouts, among other features.
+    weak! {
+        pub fn __ulock_wait2(u32, *mut c_void, u64, u64, u64) -> c_int
+    }
+
+    // These syscalls appeared with macOS 10.12.
+    weak! {
+        pub fn __ulock_wait(u32, *mut c_void, u64, u32) -> c_int
```

Comment:
Mind explaining more what the thought process of these going undetected is? Seems like a risky chance that might require a hot fix if the MAS or iOS app store scanners see through this.

In reality I think this won't be enough and we need to call `ulock` things via syscall directly instead. This means that the "fallback" Apple futex implementation is only going to be usable on macOS.

I know Electron apps on the MAS [use direct syscalls](https://github.com/electron/electron/blob/main/patches/chromium/mas_avoid_private_macos_api_usage.patch.patch#L101-L109) to emulate private functions, so this is probably what Rust should do too on macOS with precedence:
```
#define	SYS_ulock_wait     515
#define	SYS_ulock_wake     516
#define	SYS_ulock_wait2    544
```

For iOS, detect if the `os_sync_` functions are available or fallback to the non-futex implementation :/ Calling a `syscall` directly is forbidden per the iOS headers so `std` can't do the same as macOS:
```h
__WATCHOS_PROHIBITED __TVOS_PROHIBITED
__OS_AVAILABILITY_MSG(ios,deprecated=10.0,"syscall(2) is unsupported; "
    "please switch to a supported interface. For SYS_kdebug_trace use kdebug_signpost().")
__OS_AVAILABILITY_MSG(macosx,deprecated=10.12,"syscall(2) is unsupported; "
    "please switch to a supported interface. For SYS_kdebug_trace use kdebug_signpost().")
int	 syscall(int, ...);
```

Reply:
This is a "spirit of the law" kind of situation. Apple understandably wants App Store software to continue working in newer versions, so they prohibit private API use, as they want to be able to remove or change that API at will. By only using the private API as fallback and not linking it directly, we fulfil this wish. If we directly linked the private API instead, we'd get dynamic linker errors if they remove it, so we can't do that. 

On the other hand, we break the "letter of the law" as we are using private API. In my opinion, this is totally fine and justified, but of course, this is only my interpretation. It would be great if someone from Apple could look over this, just so that we know that they are aware of this. My past attempt at reaching out to them for this has been unsuccessful.

Reply:
Would you consider the solution of `atomic-wait`? They use libc++.

https://github.com/m-ou-se/atomic-wait/blob/main/src/macos.rs

Reply:
But these only exist starting with macOS 11, which is way above our minimum of 10.12, so the ulock fallback would be needed regardless.

Reply:
This sets precedent for using private APIs in the standard library (our current use of `weak!` is only used for symbols that are available in newer versions), which I think is a fairly big deal, and should perhaps be discussed more broadly (in a separate issue? (that could be FCP'ed?)).

Somewhat related is https://github.com/rust-lang/rust/issues/114186. Tagging in particular @thomcc and @workingjubilee, as they seem to have been involved in this kind of stuff before?

---

At the very least, we should use the `dlsym!` macro explicitly here, to make it very clear that we're _not_ weakly linking the symbol, but actually loading it at runtime, and thereby evading the App Store's checks ([`dlsym` itself seems to be allowed](https://forums.developer.apple.com/forums/thread/130502)). I _believe_ this is sufficient such that we do not need to do raw syscalls (which I remember to have caused problems for Go in the past since as the ABI isn't stable (?)).

Reply:
Yeah, we shouldn't be calling this directly, it'll trip the app store static analysis.